sbt-scala-native/src/main/scala/scala/scalanative/sbtplugin/ScalaNativePluginInternal.scala

```diff
@@ -284,6 +284,12 @@ object ScalaNativePluginInternal {
       "org.scala-native" %%% "javalib"   % nativeVersion,
       "org.scala-native" %%% "scalalib"  % nativeVersion
     ),
+    javacOptions ++= Seq("-source", "1.8", "-target", "1.8"),
+    initialize := {
+      val res = initialize.value
+      if (sys.props("java.specification.version") != "1.8")
+        sys.error("Java 8 is required for this project.")
+    },
```

Comment:
Java 9 and newer should be fine too, it's only blocked by the availability of the new NIO APIs.